File: search_sort/sort/quickSort/quickSort.py
def quickSort(aList):

    _quickSort(aList, 0, len(aList))


def _quickSort(aList, first, last):

    if first >= last:

        return

    else:

        pivot = partition(aList, first, last)
        # print pivot

        _quickSort(aList, first, pivot - 1)
        _quickSort(aList, pivot, last)


def partition(aList, first, last):
    pivot = aList[first]
    print 'pivot'
    print pivot
    lessIndex = first + 1

    for moreIndex in range(lessIndex, last):

        if aList[moreIndex] <= pivot:

            swap(aList, lessIndex, moreIndex)

            lessIndex += 1

    swap(aList, first, lessIndex - 1)

    print aList

    return lessIndex


def swap(aList, x, y):
    temp = aList[y]
    aList[y] = aList[x]
    aList[x] = temp


test1 = [4, 6, 13, 6, 2, 7, 346, 15, 64, 246, 2, 6, 64]
test2 = [5, 3, 135, 3, 6]
test3 = [1]

quickSort(test1)
# quickSort(test2)
# quickSort(test3)
# print quickSort(test1)
# print quickSort(test2)
# print quickSort(test3)

print 'lol'
print test1
# print test2
# print test3

File: search_sort/sort/quickSort/quickSort.js
function quickSort(arr) {
  _quickSort(arr, 0, arr.length);
}

function _quickSort(arr, first, last) {

  if (first >= last) {

    return;

  } else {
    var pivot = partition(arr, first, last);

    _quickSort(arr, first, pivot - 1);
    _quickSort(arr, pivot, last);
  }
}

function partition(arr, first, last) {
  var pivot = arr[first];

  //console.log('! ' + pivot);

  var lessIndex = first + 1;

  for (var moreIndex = lessIndex; moreIndex < last; moreIndex++) {

    if (arr[moreIndex] <= pivot) {

      swap(arr, moreIndex, lessIndex);
      lessIndex++;

    }

  }

  swap(arr, first, lessIndex - 1)

  //console.log(arr);

  return lessIndex;

}

function swap(arr, x, y) {
  var temp = arr[y];
  arr[y] = arr[x];
  arr[x] = temp;
}

test1 = [4, 6, 13, 5, 2, 7, 34, 12, 64, 1, 71, 6]
test2 = [5, 3, 135, 3, 6]
test3 = [1]
test4 = [4, 6, 13, 6, 2, 7, 346, 15, 64, 246, 2, 6, 64]

quickSort(test1)
quickSort(test2)
quickSort(test3)
quickSort(test4)
  // console.log(quickSort(test1))
  // console.log(quickSort(test2))
  // console.log(quickSort(test3))

console.log(test1)
console.log(test2)
console.log(test3)
console.log(test4)

//quickSort(test2)